### PR TITLE
Basic scala scaffolding

### PR DESCRIPTION
**Pull request by greggdonovan**
(Any real logic is stripped out for now to keep the commit size reasonable.)

- fix a typo in java.bzl and enable it in the prelude
- import `rules_scala` in the `WORKSPACE`
- add scala-libary and scalatest dependencies
- test and build with a persistent scalac worker
- add `scala_library` and `scala_test` to the prelude
- add current workarounds for IntelliJ support:
  - add `scala_import` and  by @ittaiz
  - add `scala_maven_jar` and `declare_maven` wrapper by @kevingessner

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: tools/build_rules/prelude_bazel [hunk 1]
```diff
@@ -1,2 +1,2 @@
 load("@//tools/build_rules:java.bzl", "java_library")
-
+load("@io_bazel_rules_scala//scala:scala.bzl", "scala_library", "scala_test")
```
Comment (cgrushko) [inline tools/build_rules/prelude_bazel]: mm... how do you feel about loading this only in BUILD files that need it?

File: tools/build_rules/workspace.bzl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+scala_maven_jar = repository_rule(
+    implementation = _scala_maven_jar_impl,
+    attrs = {
+        "jar": attr.label(),
+    },
+    local = True,
+)
+
+def declare_maven(hash):
```
Comment (cgrushko) [inline tools/build_rules/workspace.bzl]: this file is confusing to me :)
I'd expect scala_maven_jar to be in its own `scala_maven_jar.bzl` file, and the `declare_maven` to be defined in //:WORKSPACE and be passed into maven_repositories() ?

**Review by cgrushko — COMMENTED** [review]
Awesome!!

Google's style says that Skylark code should be put in //tools/bazel_defs instead of //tools/bazel_rules, presumably because it describes it better. I don't have the full context for that decision, but I prefer to follow it if it's not a huge hassle.

File: tools/build_rules/prelude_bazel (hunk 1 above)
Comment (greggdonovan) [reply tools/build_rules/prelude_bazel]: Right, that makes sense given that Scala is only a small percentage of the project.

**Comment by greggdonovan** [comment]
Sure, I'll move the Skylark into `//tools/bazel_defs`. I also noticed that I forgot the copyright headers. I'll update the PR.

**Comment by greggdonovan** [comment]
Should we move `//tools/build_rules/java.bzl` into `//tools/bazel_defs` as well?

**Comment by cgrushko** [comment]
Moving `//tools/build_rules/java.bzl` also makes sense. Thanks!

File: tools/build_rules/prelude_bazel (hunk 1 above)
Comment (greggdonovan) [reply tools/build_rules/prelude_bazel]: This is a bit tricker than I thought. bazel-deps will not generate the load statements needed for scala_library unless they are specified manually in the `buildHeader`. If you do that, they get added to every BUILD file in `thirdparty/jvm`.

Should we add an option to bazel-deps to autogenerate the load statements based on the types of artifacts that are being written to the BUILD file?

File: tools/build_rules/workspace.bzl (hunk 2 above)
Comment (greggdonovan) [reply tools/build_rules/workspace.bzl]: > I'd expect scala_maven_jar to be in its own scala_maven_jar.bzl file

That's a good idea. Fixing.

> declare_maven to be defined in //:WORKSPACE and be passed into maven_repositories()
 
Oops. It should be. I broke that when splitting up the commits. Fixing.

File: tools/bazel_defs/workspace.bzl [hunk 3]
```diff
@@ -0,0 +1,63 @@
+load('//tools/bazel_defs:scala_maven_jar.bzl', 'scala_maven_jar')
+
+def declare_maven(hash):
```
Comment (cgrushko) [inline tools/bazel_defs/workspace.bzl]: mmm.. rename the file to declare_maven?

**Review by cgrushko — APPROVED** [review]
Re loading `scala_library` only when needed -
1. GitHub ate this comment so good thing I read the email. I'm seriously thinking about switching to Gerrit. It's also horrible but at least I don't miss comments there...
2. Ah, I see you already filed an issue :) thanks! So let's load the Scala rules in the prelude but have a  `# TODO(https://github.com/johnynek/bazel-deps/issues/86): Remove once bazel-deps generates load() statements where needed`.

Thanks! This is awesome :D 
Feel free to merge after addressing this comment and the one about renaming workspace.bzl.

File: tools/bazel_defs/workspace.bzl (hunk 3 above)
Comment (greggdonovan) [reply tools/bazel_defs/workspace.bzl]: Good idea, will do.

**Comment by greggdonovan** [comment]
test this please

**Comment by greggdonovan** [comment]
retest this please

**Comment by greggdonovan** [comment]
@cgrushko @petroseskinder Can you kick off a CI run? Reading the [CI docs](https://github.com/bazelbuild/continuous-integration/blob/master/docs/user.md#github-pull-requests), it looks like I have to be in the [bazelbuild org](https://github.com/orgs/bazelbuild/people) to kick them off myself.

**Comment by cgrushko** [comment]
retest this please

**Pull request merged**